### PR TITLE
Enhance the Http Firehose to work with URIs requiring basic authentication

### PR DESCRIPTION
**Pull request by capistrant**
Related to #7122 

This PR would allow clients submitting API requests for `index` or `index_parallel` tasks that utilize the HTTP firehose to optionally specify username/password credentials within the `ioConfig["firehose"]` json block of the spec. The credentials will be used to authenticate against URIs which require basic authentication in order to successfully submit a GET request to said URIs.

The change would not break any existing client code that uses the HTTP firehose because omitting the new JSON key:value pairs from relevant JSON requests will result in the code acting as it did before this change:

- If the credentials are specified, the `HttpFireHoseFactory` creates and adds a basic authentication header for HTTP requests.
- If the credentials are omitted from the spec, the `HttpFireHoseFactory` does not create a basic authentication header for HTTP requests.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java [hunk 1]
```diff
@@ -46,6 +47,8 @@
   private static final Logger log = new Logger(HttpFirehoseFactory.class);
   private final List<URI> uris;
   private final boolean supportContentRange;
+  private final String httpAuthenticationUsername;
+  private final String httpAuthenticationPassword;
```
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java]: This will be printed in the task log if it's string. Please use `PasswordProvider` to make it configurable.

File: server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java [hunk 2]
```diff
@@ -81,7 +100,13 @@ public HttpFirehoseFactory(
   @Override
   protected InputStream openObjectStream(URI object) throws IOException
   {
-    return object.toURL().openConnection().getInputStream();
+    URLConnection urlConnection = object.toURL().openConnection();
+    if (!"".equals(httpAuthenticationUsername) && !"".equals(httpAuthenticationPassword)) {
```
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java]: Please add this to another method `openObjectStream(URI object, long start)` too.

Comment (jihoonson) [reply server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java]: Also, please add a unit test which checks that the request property is set properly. You probably need to extract a common part to a new method which opens and annotates a connection for testing.

**Comment by jihoonson** [comment]
@capistrant thanks! I left some comments.

**Comment by jihoonson** [comment]
Also, please check this CI failure: https://travis-ci.org/apache/incubator-druid/jobs/498858420. It looks legit.

File: server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java (hunk 1 above)
Comment (capistrant) [reply server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java]: Thank you for the comment. Could you elaborate a bit on what you mean by configurable? Once wrapped in a PasswordProvider, won't the content still need to be serialized in plaintext for it to be passed between services in the cluster (overlord --> worker)? It is not clear to me how we could selectively mask the password just by encapsulating it within a password provider. I've been trying to run a search on the codebase to find a similar situation but have come up empty so far.

Any tips are appreciated!

**Comment by capistrant** [comment]
@jihoonson I have had time to go back and make changes to the PR. Per my comment above regarding the PasswordProvider, it was not obvious to me that there was some native way for me to use a password provider to prevent jackson from serializing the password field into json written to select locations (the payload and log were the two of note). Due to that, my solution to remove the password from persisted locations was to add a MixIn to the JsonMapper objects responsible for serialization. When doing this, I only wanted to add the MixIn at very select times so I created new object mappers using the existing one as template and then added the MixIn. I choose to do this because changing the Object mapper across the project means the MixIn was persisted to the object mapper and used for serialization at times when I did not want it to be.

I'm not sure if this the best approach to fixing the Jackson serialization problem that led to the security vulnerability. If I overlooked some better approach, I am happy to amend.

**Comment by jihoonson** [comment]
Oh sorry, I missed your question. I'll take another look soon. Thank you for the update!

File: server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java [hunk 3]
```diff
@@ -54,7 +59,9 @@ public HttpFirehoseFactory(
       @JsonProperty("maxFetchCapacityBytes") Long maxFetchCapacityBytes,
       @JsonProperty("prefetchTriggerBytes") Long prefetchTriggerBytes,
       @JsonProperty("fetchTimeout") Long fetchTimeout,
-      @JsonProperty("maxFetchRetry") Integer maxFetchRetry
+      @JsonProperty("maxFetchRetry") Integer maxFetchRetry,
+      @JsonProperty("httpAuthenticationUsername") String httpAuthenticationUsername,
+      @JsonProperty("httpAuthenticationPassword") DefaultPasswordProvider httpAuthenticationPasswordProvider
```
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java]: Any reason to allow only `DefaultPasswordProvider`? The intention of using `PasswordProvider` is to provide `DefaultPasswordProvider` by default, but users can use `EnvironmentVariablePasswordProvider` or their own custom passwordProvider if they want to hide password information from logs.

File: server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java [hunk 4]
```diff
@@ -46,6 +49,8 @@
   private static final Logger log = new Logger(HttpFirehoseFactory.class);
   private final List<URI> uris;
   private final boolean supportContentRange;
+  private final String httpAuthenticationUsername;
```
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java]: Please annotate `@Nullable`.

File: server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java [hunk 5]
```diff
@@ -54,7 +59,9 @@ public HttpFirehoseFactory(
       @JsonProperty("maxFetchCapacityBytes") Long maxFetchCapacityBytes,
       @JsonProperty("prefetchTriggerBytes") Long prefetchTriggerBytes,
       @JsonProperty("fetchTimeout") Long fetchTimeout,
-      @JsonProperty("maxFetchRetry") Integer maxFetchRetry
+      @JsonProperty("maxFetchRetry") Integer maxFetchRetry,
+      @JsonProperty("httpAuthenticationUsername") String httpAuthenticationUsername,
```
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java]: Please annotate `@Nullable`.

File: server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java [hunk 6]
```diff
@@ -64,6 +71,20 @@ public HttpFirehoseFactory(
     final URLConnection connection = uris.get(0).toURL().openConnection();
     final String acceptRanges = connection.getHeaderField(HttpHeaders.ACCEPT_RANGES);
     this.supportContentRange = acceptRanges != null && "bytes".equalsIgnoreCase(acceptRanges);
+    this.httpAuthenticationUsername = httpAuthenticationUsername;
+    this.httpAuthenticationPasswordProvider = httpAuthenticationPasswordProvider == null ? new DefaultPasswordProvider("") : httpAuthenticationPasswordProvider;
```
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java]: Should this have a default value? Or is it fine to set it to null if the parameter is null?

File: server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java [hunk 7]
```diff
@@ -161,11 +189,25 @@ public int hashCode()
...
     catch (IOException e) {
       throw new RuntimeException(e);
     }
   }
+
+  @VisibleForTesting
+  URLConnection openURLConnection(URI object) throws IOException
+  {
+    URLConnection urlConnection = object.toURL().openConnection();
+    if (!"".equals(httpAuthenticationUsername) && !"".equals(httpAuthenticationPasswordProvider.getPassword())) {
```
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java]: `httpAuthenticationUsername` can be null. Should use `Strings.isNullOrEmpty()`.

File: indexing-service/src/main/java/org/apache/druid/indexing/worker/executor/ExecutorLifecycle.java [hunk 8]
```diff
@@ -80,7 +83,7 @@ public ExecutorLifecycle(
     this.taskConfig = taskConfig;
     this.taskActionClientFactory = taskActionClientFactory;
     this.taskRunner = taskRunner;
-    this.jsonMapper = jsonMapper;
+    this.jsonMapper = new DefaultObjectMapper((DefaultObjectMapper) jsonMapper).addMixIn(PasswordProvider.class, PasswordProviderRedactionMixIn.class);
```
Comment (jihoonson) [inline indexing-service/src/main/java/org/apache/druid/indexing/worker/executor/ExecutorLifecycle.java]: Would you please explain why this change is needed? `KinesisIndexTask` is already using `PasswordProvider`, so it should work without this change.

File: server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java [hunk 9]
```diff
@@ -82,7 +83,8 @@ public SQLMetadataStorageActionHandler(
   )
   {
     this.connector = connector;
-    this.jsonMapper = jsonMapper;
+    //noinspection UnnecessaryFullyQualifiedName
+    this.jsonMapper = new DefaultObjectMapper((DefaultObjectMapper) jsonMapper).addMixIn(org.apache.druid.metadata.PasswordProvider.class, org.apache.druid.metadata.PasswordProviderRedactionMixIn.class);
```
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java]: Same here. Would you please explain why this change is needed? `KinesisIndexTask` is already using `PasswordProvider`, so it should work without this change.

File: server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java (hunk 3 above)
Comment (jihoonson) [reply server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java]: Also please add `@Nullable`.

File: indexing-service/src/main/java/org/apache/druid/indexing/worker/executor/ExecutorLifecycle.java (hunk 8 above)
Comment (capistrant) [reply indexing-service/src/main/java/org/apache/druid/indexing/worker/executor/ExecutorLifecycle.java]: @jihoonson When I do not swap out the ObjectMapper with a new one that has the MixIn, the password is written in plaintext to the metastore and the task log file. Both of those instances serialize the object to JSON using the existing ObjectMapper before writing, so I don't think that the ObjectMapper is configured to recognize and omit the password as needed. Also, since the way the indexing service works requires the Serialization to selectively include the password in the json (when the overlord is assigning a remote worker), changing the default object mapper to handle this new block would not work either. That is why I felt the need to use a MixIn

File: server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java (hunk 3 above)
Comment (capistrant) [reply server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java]: Thank you for calling this out. Now that I understand PasswordProvider more I get why using the vanilla `PasswordProvider` is best. I can see the environment variable being useful for teams who don't want to be sending passwords in the JSON

File: docs/content/ingestion/firehose.md [hunk 10]
```diff
@@ -74,6 +74,28 @@ A sample http firehose spec is shown below:
 }
 ```
 
+The below configurations can be optionally used if the URIs specified in the spec require a Basic Authentication Header.
+Omitting these fields from your spec will result in HTTP requests with no Basic Authentication Header.
+
+|property|description|default|
+|--------|-----------|-------|
+|httpAuthenticationUsername|Username to use for authentication with specified URIs|None|
+|httpAuthenticationPassword|json object with a field called password containing the password to use for authentication with specified URIs|None|
```
[the remark this answers is not in the document]
Comment (capistrant) [reply docs/content/ingestion/firehose.md]: Regarding, the json -> JSON. The rest of the page was using json lowercase so I kept it consistent. Should I change the whole page?

[the remark this answers is not in the document]
Comment (jihoonson) [reply docs/content/ingestion/firehose.md]: Hmm, ok then. Maybe better to change them in a separate PR.

File: server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java (hunk 6 above)
Comment (jihoonson) [reply server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java]: Would you please check this comment?

File: indexing-service/src/main/java/org/apache/druid/indexing/worker/executor/ExecutorLifecycle.java (hunk 8 above)
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/worker/executor/ExecutorLifecycle.java]: I see. It sounds good to me. Would you please this to the javadoc of `PasswordProviderRedactionMixIn`?

File: server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java (hunk 6 above)
Comment (capistrant) [reply server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java]: @jihoonson I defaulted to a Default PasswordProvider with empty string password in order to simplify the logic on line 213 (https://github.com/apache/incubator-druid/pull/7145/files#diff-4adc70bca196e07dfd5195379218c9e6R213). If we just leave with null I suppose we could change this logic to not throw NullPointerException. Do you think letting it would be null and changing the logic on 213 would be a better choice?

Comment (jihoonson) [reply server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java]: I see. I think it's better to leave it to null because the default value would appear in the log which I think can make users confused by showing a password provider which is not set by users. What do you think?

File: indexing-service/src/main/java/org/apache/druid/indexing/worker/executor/ExecutorLifecycle.java (hunk 8 above)
Comment (capistrant) [reply indexing-service/src/main/java/org/apache/druid/indexing/worker/executor/ExecutorLifecycle.java]: @jihoonson Sorry, but I'm unclear about what change you are requesting in the javadoc. Could you elaborate?

Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/worker/executor/ExecutorLifecycle.java]: Ah I meant it would be nice if you can add javadoc for `PasswordProviderRedactionMixIn` explaining why that class exists and how it's being used.

File: docs/content/ingestion/firehose.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (capistrant) [reply docs/content/ingestion/firehose.md]: Reverted back to have whole file be same with lowercase

File: server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java (hunk 6 above)
Comment (capistrant) [reply server/src/main/java/org/apache/druid/segment/realtime/firehose/HttpFirehoseFactory.java]: I agree now that I think about it. Have made the change in latest commit.

File: indexing-service/src/main/java/org/apache/druid/indexing/worker/executor/ExecutorLifecycle.java (hunk 8 above)
Comment (capistrant) [reply indexing-service/src/main/java/org/apache/druid/indexing/worker/executor/ExecutorLifecycle.java]: Sure, I added a Javadoc. Not sure if there is an expected pattern for this Project. I wasn't able to find too much consistency across the project. Will be happy to get in line if there is an expected standard out there

File: server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java (hunk 9 above)
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java]: How about fixing the constructor of `DefaultObjectMapper` to accept `ObjectMapper` instead of `DefaultObjectMapper`. I think it's better because `jsonMapper` doesn't have to be `DefaultObjectMapper`.

Comment (capistrant) [reply server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java]: @jihoonson Oh yes, that makes sense. After checking out the spec for ObjectMapper I think using copy() is the correct way to go with this. It is in the latest commit, let me know what you think about that pattern for this

File: server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java [hunk 11]
```diff
@@ -82,7 +82,9 @@ public SQLMetadataStorageActionHandler(
   )
   {
     this.connector = connector;
-    this.jsonMapper = jsonMapper;
+    //noinspection UnnecessaryFullyQualifiedName
+    this.jsonMapper = jsonMapper.copy().addMixIn(org.apache.druid.metadata.PasswordProvider.class,
```
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java]: Oh, I missed this part. Would you tell me why we should use fully qualified name here?

Comment (capistrant) [reply server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java]: PasswordProvider is in the core module, but shares the same package name. This introduced some issues importing it into the server module. If I use the import statement at the top of the file I get hit with an unused import issue, and If I don't import, the IDE will figure out how to resolve but at runtime you get a class not found error. I decided on the fully qualified name with the noinspection flag. At the same time I'm not a Java veteran so I don't know what the standard would be here

Comment (jihoonson) [reply server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java]: I see. I think it's fine, but would you please add a comment about why we use the fully qualified name here?

Comment (capistrant) [reply server/src/main/java/org/apache/druid/metadata/SQLMetadataStorageActionHandler.java]: Pushed a comment just now

**Comment by capistrant** [comment]
@jihoonson Thank you for working through the changes with me during review for this change. Much appreciated!

**Comment by jihoonson** [comment]
Thank you for implementing a cool feature!
